drivers/bluetooth/hci/rpmsg.c

```diff
@@ -0,0 +1,426 @@
+	if (!ipm_rx_handle) {
+		BT_ERR("Could not get RX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_register_callback(ipm_rx_handle, ipm_callback, NULL);
+
+	/* Virtqueue setup */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[0]) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[0]");
+		return -ENOMEM;
+	}
+
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[1]) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[1]");
+		return -ENOMEM;
+	}
+
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.role = RPMSG_MASTER;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	vdev.vrings_info = &rvrings[0];
+
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	err = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+	if (err) {
+		BT_ERR("rpmsg_init_vdev failed %d", err);
+		return err;
+	}
+
+	/* Since we are using name service, we need to wait for a response
+	 * from NS setup and than we need to process it
+	 */
+	virtqueue_notification(vq[0]);
+
+	/* Wait til nameservice ep is setup */
+	k_sem_take(&sync_sem, K_FOREVER);
+
+	return 0;
+}
+
+static inline void bt_buf_set_prio(struct net_buf *buf, bool prio)
+{
+	*((u8_t *)net_buf_user_data(buf) + 1) = prio;
+}
+
+static inline bool bt_buf_get_prio(struct net_buf *buf)
+{
+	return (bool)(*((u8_t *)net_buf_user_data(buf) + 1));
+}
+
+static struct net_buf *bt_rpmsg_evt_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_evt_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enought data for event header");
+		return NULL;
+	}
+
+	memcpy((void *)&hdr, data, sizeof(hdr));
+	data += sizeof(hdr);
+	remaining -= sizeof(hdr);
+
+	if (remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+		return NULL;
+	}
+	remaining = hdr.len;
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, false, K_NO_WAIT);
+	if (!buf) {
+		BT_ERR("No available event buffers!");
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	bt_buf_set_prio(buf, bt_hci_evt_is_prio(hdr.evt));
+
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_rpmsg_acl_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enought data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+		bt_buf_set_prio(buf, false);
+	} else {
+		BT_ERR("No available ACL buffers!");
+	}
+
+	if (remaining != hdr.len) {
```

Comment:
`remaining != sys_le16_to_cpu(hdr.len)`. That said, you'd then do the conversion twice, so if you want to avoid that some redesign is needed. Note that for ACL the length header is 16 bits, but for HCI events it's 8 bits (so no conversion necessary there)

Reply:
Actually, the line of code with the second conversion is not needed